File: src/leibniz.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;

public class leibniz {
    private static double calculate(int rounds) {
        rounds += 2;

        double pi = 1;
        double x = 1;

        for (int i = 2; i < rounds; i++) {
            x *= -1;
            pi += (x / (2 * i - 1));
        }

        return pi * 4;
    }

    public static void main(String[] args) throws IOException {
        int rounds = Integer.parseInt(Files.readString(Path.of("rounds.txt")));

        for (int i = 0; i < 10; i++) {
            long start = System.nanoTime();
            double pi = calculate(rounds);
            long end = System.nanoTime();
    
            double time = (end - start) / 1000_000.0;
            System.out.printf("Warmup Iteration %d (%.3fms): %.16f\n", i, time, pi);
        }
    
        double totalTime = 0;
    
        for (int i = 0; i < 10; i++) {
            long start = System.nanoTime();
            double pi = calculate(rounds);
            long end = System.nanoTime();
    
            double time = (end - start) / 1000_000.0;
            System.out.printf("Iteration %d (%.3fms): %.16f\n", i, time, pi);

            totalTime += time;
        }
        System.out.printf("Average: %fms\n", totalTime / 10);
    }
}

File: src/leibniz.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"
)

func calculate(rounds int) float64 {
	rounds += 2

	pi := 1.0
	x := 1.0

	for i := 2; i < rounds; i++ {
		x *= -1
		pi += x / float64(2*i-1)
	}

	return pi
}

func main() {
	file, err := os.ReadFile("rounds.txt")
	if err != nil {
		fmt.Print(err)
	}

	rounds, _ := strconv.Atoi(strings.TrimSpace(string(file)))

	for i := 0; i < 10; i++ {
		start := time.Now()
		pi := calculate(rounds)

		t := float64(time.Since(start).Nanoseconds()) / 1000_000

		fmt.Printf("Warmup Iteration %d (%.3f ms): %.16f\n", i, t, pi)
	}

	totalTime := 0.0

	for i := 0; i < 10; i++ {
		start := time.Now()
		pi := calculate(rounds)

		t := float64(time.Since(start).Nanoseconds()) / 1000_000

		fmt.Printf("Iteration %d (%.3f ms): %.16f\n", i, t, pi)

		totalTime += t
	}

	fmt.Printf("Average: %fms\n", totalTime/10)
}
